File: lib/domain/entities/transform.ts
import { Vector3 } from "./vector3"

export class Transform {
  position: Vector3
  rotation: Vector3
  scale: Vector3

  constructor(
    position: Vector3 = new Vector3(),
    rotation: Vector3 = new Vector3(),
    scale: Vector3 = new Vector3(1, 1, 1),
  ) {
    this.position = position
    this.rotation = rotation
    this.scale = scale
  }

  clone(): Transform {
    return new Transform(this.position.clone(), this.rotation.clone(), this.scale.clone())
  }

  toJSON() {
    return {
      position: this.position.toJSON(),
      rotation: this.rotation.toJSON(),
      scale: this.scale.toJSON(),
    }
  }

  static fromJSON(data: any): Transform {
    return new Transform(Vector3.fromJSON(data.position), Vector3.fromJSON(data.rotation), Vector3.fromJSON(data.scale))
  }
}

File: lib/domain/scene/transform.dart
import 'package:json_annotation/json_annotation.dart';

part 'transform.g.dart';

@JsonSerializable()
class Vec3 {
  final double x;
  final double y;
  final double z;

  Vec3({required this.x, required this.y, required this.z});

  factory Vec3.fromJson(Map<String, dynamic> json) => _$Vec3FromJson(json);
  Map<String, dynamic> toJson() => _$Vec3ToJson(this);
}

@JsonSerializable()
class Transform {
  final Vec3 position;
  final Vec3 rotation;
  final Vec3 scale;

  Transform({required this.position, required this.rotation, required this.scale});

  factory Transform.fromJson(Map<String, dynamic> json) => _$TransformFromJson(json);
  Map<String, dynamic> toJson() => _$TransformToJson(this);
}
